aws/resource_aws_appmesh_virtual_service.go

```diff
@@ -0,0 +1,227 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/appmesh"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+func resourceAwsAppmeshVirtualService() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsAppmeshVirtualServiceCreate,
+		Read:   resourceAwsAppmeshVirtualServiceRead,
+		Update: resourceAwsAppmeshVirtualServiceUpdate,
+		Delete: resourceAwsAppmeshVirtualServiceDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceAwsAppmeshVirtualServiceImport,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(1, 255),
+			},
+
+			"mesh_name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(1, 255),
+			},
+
+			"spec": {
+				Type:     schema.TypeList,
+				Required: true,
+				MinItems: 1,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"provider": {
```

Comment:
Provider in the SDKs is a `union` type, which in AWS' language means only one type can be provided (in this case, either VirtualNode OR VirtualRouter). Just curious how this plays with Terraform? Is it validated here or do you rely on the App Mesh control plane to invalidate when values are provided for both?

Reply:
* RE https://github.com/terraform-providers/terraform-provider-aws/pull/7858#discussion_r263634391: Once I add listeners then I can move `Removed` down to `service_names` and change the message to mention the virtual service resource.

* RE https://github.com/terraform-providers/terraform-provider-aws/pull/7858#discussion_r263635145: Yes, we can do that with a `ConflictsWith` attribute